keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1108 @@
+and UI interaction.
+
+### API
+The go code below describes the Application CRD. 
+
+```go
+
+// ApplicationStatus contains the status of an Application. 
+type ApplicationStatus struct {
+    // ObservedGeneration is used by the Application Controller to report the last Generation of an Application 
+    // that it has observed.
+    ObservedGeneration int64 `json:"observedGeneration,omitempty" protobuf:"int64,1,opt,name=observedGeneration"`
+    
+    // Installed is used by the Application Controller to report the installed Components and Dependencies of  
+    // an Application.
+    Installed []string `json:"installed,omitempty" protobuf:"bytes,2,opt,name=installed"`
+    
+    // Ready is used to report the readiness of an Application. If an Application contains an ApplicationHealthCheck, 
+    // this field corresponds directly to the Readiness of the Pod generated from that health check. If the 
+    // Application does not contain an ApplicationHealthCheck, this field is set to Unknown.
+    Ready ConditionStatus json:`"ready,omitempty" protobuf:"bytes,3,opt,name=ready"`
+}
+
+// ApplicationSpec is the specification object for an Application. It specifies the Application's components, 
+// dependencies, and an optional health check. Application components can be retrieved using the Applications 
+// spec.selector to select the resources that match from the components GroupKind at the client's desired version.
+type ApplicationSpec struct {
+    // Type is the type of the application (e.g. WordPress, MySQL, Cassandra).
+    Type string `json:"type" protobuf:"bytes,1,name=type"`
+    
+    // Components is a map of the applications components to the kinds created by the application (e.g. Pods, Services, 
+    // Deployments, CRDS, etc). The keys of the Components map are the identifiers of the components. These may not 
+    // be the names of the components. For example, two instances of an Application of the same type may use components 
+    // with same identifier key, but the names of the components must differ in order to ensure uniqueness in the 
+    // namespace. The values in the Components map are corresponding GroupKinds of the Application's components. These 
+    // may be used, in conjunction with the Selector, to retrieve an Application's components at the client's desired 
+    // version.
+    Components map [string]metav1.GroupKind `json:"components,omitempty" protobuf:"bytes,2,opt,name=components"`
+    
+    // Dependencies is a list of other Applications that this Application depends on.
+    Dependencies []sting `json:"dependencies" protobuf:"bytes,3,opt,name=dependencies"`
+    
+    // Selector is a label query over kinds that created by the application. It must match the component objects' labels.
+    // More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+    Selector *metav1.LabelSelector `json:"selector" protobuf:"bytes,4,opt,name=selector"`
```

Comment:
I might have missed the explanation somewhere in the document, but why the extra-redundancy here? Is the author supposed to make sure that the components have these selectors? what happens if it doesn't? 

If this isn't pure redundancy, why should I include some items in components and target some others with the selector?